data-prepper-core/src/main/java/org/opensearch/dataprepper/pipeline/ProcessWorker.java

```diff
@@ -40,41 +40,65 @@ public ProcessWorker(
     @Override
     public void run() {
         try {
-            do {
-                final Map.Entry<Collection, CheckpointState> readResult = readBuffer.read(pipeline.getReadBatchTimeoutInMillis());
-                Collection records = readResult.getKey();
-                final CheckpointState checkpointState = readResult.getValue();
-                //TODO Hacky way to avoid logging continuously - Will be removed as part of metrics implementation
-                if (records.isEmpty()) {
-                    if(!isEmptyRecordsLogged) {
-                        LOG.info(" {} Worker: No records received from buffer", pipeline.getName());
-                        isEmptyRecordsLogged = true;
-                    }
-                } else {
-                    LOG.info(" {} Worker: Processing {} records from buffer", pipeline.getName(), records.size());
-                }
-                //Should Empty list from buffer should be sent to the processors? For now sending as the Stateful processors expects it.
-                for (final Processor processor : processors) {
-                    records = processor.execute(records);
-                }
-                if (!records.isEmpty()) {
-                    postToSink(records);
-                }
-                // Checkpoint the current batch read from the buffer after being processed by processors and sinks.
-                readBuffer.checkpoint(checkpointState);
-            } while (!shouldStop());
+            // Phase 1 - execute until stop requested
+            while (!pipeline.isStopRequested()) {
+                doRun();
+            }
+            LOG.info("Processor shutdown phase 1 complete.");
+
+            // Phase 2 - execute until buffers are empty
+            LOG.info("Beginning processor shutdown phase 2, iterating until buffers empty.");
+            while (!readBuffer.isEmpty()) {
+                doRun();
+            }
+            LOG.info("Processor shutdown phase 2 complete.");
+
+            // Phase 3 - execute until peer forwarder drain period expires (best effort to process all peer forwarder data)
+            final long drainTimeoutExpiration = System.currentTimeMillis() + pipeline.getPeerForwarderDrainTimeout().toMillis();
```

Comment:
I think this part could be part of the MultiBuffer (if we viewed it more as a peer-forward-aware buffer).